pull/github.go

```diff
@@ -798,13 +798,47 @@ func (ghc *GitHubContext) loadPagedData() error {
 func (ghc *GitHubContext) loadCommits() ([]*Commit, error) {
 	log := zerolog.Ctx(ghc.ctx)
 
+	// GitHub's API is eventually consistent for commit pushed dates. We can
+	// sometimes get around this by using alternate APIs (see notes in
+	// loadCommitsOnce), but for some merge commits, that still fails. The only
+	// fix I know of in this case is to wait and try again.
+	//
+	// Retry with exponential backoff until it works or we hit the max total
+	// latency to avoid users thinking the bot got stuck or dropped an event.
+	const baseDelay = 1 * time.Second
+	const maxLatency = 20 * time.Second
+
+	start := time.Now()
+	for delay := baseDelay; true; delay *= 2 {
+		head, commits, err := ghc.loadCommitsOnce()
+		if err != nil {
+			return nil, err
+		}
+		if head.PushedAt != nil {
+			return commits, nil
+		}
+
+		if time.Since(start)+delay >= maxLatency {
+			break
+		}
+		log.Debug().Dur("delay", delay).Str("sha", ghc.pr.HeadRefOID).Msg("Head commit is missing pushed data, sleeping and trying again")
+		time.Sleep(delay)
+	}
+
+	return nil, &TemporaryError{
+		fmt.Sprintf("Commit %.10s does not have a pushed date. This is usually caused by delays in the GitHub API", ghc.pr.HeadRefOID),
+	}
+}
+
+func (ghc *GitHubContext) loadCommitsOnce() (head *Commit, commits []*Commit, err error) {
```

Comment:
I'll assume that this is only called one function above so we can have it do whatever.